exercises/src/main/scala/fpinscala/monads/Monad.scala

```diff
@@ -34,25 +34,49 @@ trait Monad[M[_]] extends Functor[M] {
   def map2[A,B,C](ma: M[A], mb: M[B])(f: (A, B) => C): M[C] =
     flatMap(ma)(a => map(mb)(b => f(a, b)))
 
-  def sequence[A](lma: List[M[A]]): M[List[A]] = ???
-
-  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] = ???
-
-  def replicateM[A](n: Int, ma: M[A]): M[List[A]] = ???
-
-  def compose[A,B,C](f: A => M[B], g: B => M[C]): A => M[C] = ???
+  def sequence[A](lma: List[M[A]]): M[List[A]] =
+    lma.foldRight(unit(Nil:List[A]))((p, acc) => {
+      map2(p, acc)(_ :: _)
+    })
+
+  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] =
+    la.foldRight(unit(Nil:List[B]))((p, acc) => {
+      map2(f(p), acc)(_ :: _)
+    })
+
+  def replicateM[A](n: Int, ma: M[A]): M[List[A]] =
+    if (n <= 0)
+      unit(Nil)
+    else
+      map2(ma, replicateM(n - 1, ma))(_ :: _)
+
+  def filterM[A](ms: List[A])(f: A => M[Boolean]): M[List[A]] =
+    ms.foldRight(unit(Nil:List[A]))((a, acc) => {
+      flatMap(f(a)) { p =>
```

Comment:
Why the `flatMap` here to put it back into a monad with unit? I think you should be able to use the result of `f(a)` directly in `map2`?